prometheus_alerting_consistency.md

```diff
@@ -0,0 +1,105 @@
+
+Reserve critical level alerts only for reporting conditions that may lead to service unavailability.
+Failures of individual components should not trigger critical level alerts in general, unless they would result in that condition.
+Configure critical level alerts so they fire before the situation becomes irrecoverable.
+
+Example critical alert:
+
+```yaml
+- alert: BackendServiceStuck
+  expr: backend_state != 1
+  for: 10m
+  labels:
+    severity: critical
+  annotations:
+    summary: 'The backend service is stuck in a non-ready state'
+    description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'
+    sop_url: 'https://example.com/backend_service_stuck.asciidoc'
+```
+
+This alert fires if a backend service has *not* been ready for the last 10 minutes.
+This is a clear example of a critical issue that represents a threat to the operability of the service, and likely warrants paging someone.
+The alert has a clear summary and description annotations, and it links to a SOP with information on investigating and resolving the issue.
+
+The group of critical alerts should be small, very well-defined, highly documented, polished and with a high bar set for entry.
+
+## Warning Alerts
+
+The vast majority of alerts should use this severity.
+Issues at the warning level should be addressed in a timely manner, but don't pose an immediate threat to the operation of the service as a whole.
+
+Timeline: ~60 minutes
+
+If your alert does not meet the criteria in "Critical Alerts" above, it belongs to the warning level or lower.
+
+Use warning level alerts for reporting conditions that may lead to inability to deliver individual features of the service, but not the service as a whole.
+Most alerts are likely to be warnings.
+Configure warning level alerts so that they do not fire until components have sufficient time to try to recover from the interruption automatically.
+You should have a policy for triaging warning alerts and responding in a timely manner e.g. issues created automatically on a backlog that is being actively worked from.
+
+Example warning alert:
+
+```yaml
+- alert: ZookeeperPersistentVolumeFillingUp
+  expr: (kubelet_volume_stats_available_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"} / kubelet_volume_stats_capacity_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"} < 0.15) and predict_linear(kubelet_volume_stats_available_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"}[6h], 4 * 24 * 3600) < 0
+  for: 1h
+  labels:
+    severity: warning
+  annotations:
+    summary: 'Zookeeper PersistentVolume is filling up.'
+    description: 'Based on recent sampling, the Zookeeper PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} is expected to fill up within four days. Currently {{ $value | humanizePercentage }} is available.'
```

Comment:
Another "why do I care?" nit: what happens if the volume fills up?  If I have a bunch of fires going on, how do I rank the risk of this against those?

Reply:
Great question.
I can't think how you can answer it without getting into the specifics of a service.
Although the ability to inhibit a rule exists, so you could structure you rules so that alerts that you know are the cause of another set of alerts firing can be prioritised.
Is that worth mentioning here?

Reply:
Any further thoughts on this @wking ?

Reply:
“Storage full” is one of those bread and butter alerts which should be standardised as much as possible. I’m not such a great fan of “disk is filling” alerts as they’re hard to tune and prone to making unnecessary noise, but if anyone wants to use them (with care) the above seems fine. But it’s just another tool in the box, not necessary an essential component.

The other thing which makes me leery of “disk filling” is that during any period of abnormal operation disk usage patterns inevitably change (logs can’t be flushed from temporary storage, queued messages stack up in the spool, all that stuff) and that’s totally normal - so alerting on it as a routine matter when other stuff is on fire is likely to be a red herring.

“Disk full” (or >98% or whatevs) *is* a useful alert because it’s solidly actionable and a good starting point for both debugging and long-term remediation of an issue, gives hints on future growth and capacity planning, etc etc.